### PR TITLE
[Refinement] re-implement ecu_info.yaml and proxy_info.yaml in otaclient.configs

### PR DESCRIPTION
**Pull request by Bodong-Yang**
## Description

This PR re-implements `ecu_info.yaml` and `proxy_info.yaml` parsing logic with pydantic under `otaclient.configs` package.

With this PR, the following refinements are implemented:
1. `ecu_info` and `proxy_info` are now otaclient package scope config files, being loaded only once at otaclient starts up.
2. proper and strict checks by pydantic are applied during config files loading.
3. `ecu_info.yaml` and `proxy_info.yaml` are now parsed accordingly to the specifications.  

For more details, please check related links.

## Check list

- [x] documentations for these changes are prepared.
- [x] test file(s) that covers the change(s) is implemented.
- [x] local test is passed. 

## Major Code Changes

1. `otaclient.configs.ecu_info` is implemented with pydantic to supersede the old `otaclient.app.ecu_info` module, `ecu_info` now only loaded once on otaclient starts up.
2. `otaclient.configs.proxy_info` is implemented with pydantic to supersede the old `otaclient.app.proxy_info` module, `proxy_info` now only loaded once on otaclient starts up.
3. `BootloaderType` enum definition is moved from `otaclient.boot_control` package to `otaclient.configs.ecu_info`, and `otaclient.boot_control` now refers to the new location to use `BootloaderType` enum.
4. `otaclient.app.log_setting` now doesn't do `ecu_info.yaml` pre-loading, instead using the `otaclient.configs.ecu_info` instead. 
5. `otaclient.app` package now imports `ecu_info` and `proxy_info` from `otaclient.configs` package into `otaclient.app.configs` virtual config module, and related modules now uses `ecu_info` and `proxy_info` as package scope configs.
6. `otaclient.configs.app_cfg`: do not use `IPvAnyAddress` from pydantic to type hint IP address field anymore, instead use new defined `otaclient._utils.typing.IPAdressAny`.
7. fix up related test files according to the new changes.

## Behavior changes

Does this PR introduce behavior change(s)?

- [x] Yes, internal behaivor (will not impact user experience).

### Behavior with this PR

1. `ecu_info.yaml` and `proxy_info.yaml` now has JSON schema defined, and being checked against with when loading these configuration files on otaclient start.
8. each fields of configuration files are now checked strictly.
9. the detailed errors for a failed configuration file parsing are provided by pydantic.

## Breaking change

Does this PR introduce breaking change?
- [x] No.

## Related links & tickets
(Internal documentations)
1. [otaclient package scope configs infra](https://tier4.atlassian.net/l/cp/7X4i6uAC)
4. [ecu_info.yaml specification](https://tier4.atlassian.net/l/cp/3XKBbEpX)
5. [proxy_info.yaml specification](https://tier4.atlassian.net/l/cp/7kk1VJeH)

### CODE REVIEW

**Comment by Bodong-Yang** [comment]
@Youneko I am considering rename the gateway field in proxy_info.yaml to be gateway_otaproxy, so I will temporarily set the PR back to draft, thank you!

File: otaclient/configs/proxy_info.py [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+    enable_local_ota_proxy=True,
+)
+
+
+def parse_proxy_info(proxy_info_file: StrOrPath) -> ProxyInfo:
+    try:
+        loaded_proxy_info = yaml.safe_load(Path(proxy_info_file).read_text())
+        assert isinstance(loaded_proxy_info, dict), "not a valid yaml file"
+        _deprecation_check(loaded_proxy_info)
+        return ProxyInfo.model_validate(loaded_proxy_info, strict=True)
```
Comment (Youneko) [inline otaclient/configs/proxy_info.py:129]: [IMO] It would be useful to output the loaded proxy info to the log for analysis.

**Review by Youneko — COMMENTED** [review]
@Bodong-Yang 

I commented.
Please check them!

File: otaclient/configs/app_cfg.py [hunk 2]
```diff
@@ -299,13 +304,19 @@ class _ConfigurableConfig(BaseModel):
     """
 
     # name of OTA used temp folder
-    OTA_TMP_DNAME: str = "ota_tmp"
+    OTA_TMP_DNAME: _std_ClassVar = "ota_tmp"
```
[the remark this answers is not in the document]
Comment (Bodong-Yang) [reply otaclient/configs/app_cfg.py:307]: Ah this is unintended, I mistakenly think it is a const, but this attr is in `_ConfigurableConfig`, it should be str and configurable via environmental vars. I will fix it. 

------
For the purpose of this hack, to let IDE(I am using vscode) hint this class constant as `Literal["ota_tmp"]` without typing it as `OTA_TMP_DNAME: ClassVar[Literal"ota_tmp"] = "ota_tmp"`.
image

If typing this constant as `str`, IDE will not hint this field's value.
image

Also note that, pydantic doesn't allow `ClassVar` without parameterize at runtime, while python typing's `Literal` cannot be used without parameterizing against static type check. So typing this field as `OTA_TMP_DNAME: ClassVar[Literal]` or `OTA_TMP_DNAME: ClassVar` will not work.

File: otaclient/configs/ota_service_cfg.py [hunk 3]
```diff
@@ -29,10 +28,10 @@ class OTAServiceConfig(BaseFixedConfig):
     model_config = ConfigDict(frozen=True, validate_default=True)
 
     # used when listen_addr is not configured in ecu_info.yaml.
-    DEFAULT_SERVER_ADDRESS: Union[IPvAnyAddress, Literal["127.0.0.1"]] = "127.0.0.1"
+    DEFAULT_SERVER_ADDRESS: IPAddressAny = "127.0.0.1"
 
-    SERVER_PORT: int = Field(default=50051, ge=0, le=65535)
-    CLIENT_CALL_PORT: int = Field(default=50051, ge=0, le=65535)
+    SERVER_PORT: NetworkPort = 50051
+    CLIENT_CALL_PORT: NetworkPort = 50051
```
[the remark this answers is not in the document]
Comment (Bodong-Yang) [reply otaclient/configs/ota_service_cfg.py:34]: Yes it is possible, `SERVER_PORT` is the port local otaclient OTA grpc server listens on. `CLIENT_CALL_PORT` is the port otaclient used to contact sub ECUs OTA grpc server.

But in real world condition, we SHOULD NOT change these settings, changing these values will make maintenance difficult and possibly breaking the connection between main ECU and sub ECUs in multiple ECU environment.

File: otaclient/configs/ecu_info.py [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+        if len(self.available_ecu_ids) == 0:
+            return [self.ecu_id]
+        return self.available_ecu_ids.copy()
+
+
+# NOTE: this is backward compatibility for old x1 that doesn't have
+#       ecu_info.yaml installed.
+DEFAULT_ECU_INFO = ECUInfo(
+    format_version=1,  # current version is 1
+    ecu_id="autoware",  # should be unique for each ECU in vehicle
```
[the remark this answers is not in the document]
Comment (Bodong-Yang) [reply otaclient/configs/ecu_info.py:110]: Actually there is an implementation detail about this `unspecified` value. If the `bootloader` type is `unspecified`, otaclient will try to detect the bootloader type on itself. A.K.A, `unspecified` means involving bootloader type detecting mechanism.

------
Emmmm I think I can rename the default value to something like `auto_detect` to reduce ambiguous.

------
In the early day, `ecu_info` doesn't contain `bootloader` field, otaclient always detects the bootloader type. Later `bootloader` field is introduced and we respect the set value in `ecu_info.yaml` by default.

File: otaclient/configs/app_cfg.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (Bodong-Yang) [reply otaclient/configs/app_cfg.py:307]: The `_std_ClassVar` hack is meant to be used in class contains real constants, like this:
https://github.com/tier4/ota-client/blob/4c2b5a033ed9b9b5d9a465cc45b11d0925b70f15/otaclient/configs/app_cfg.py#L63-L69

The attrs in the final `app_cfg` will be typed in IDE as:
image

**Comment by Bodong-Yang** [comment]
@Youneko The `proxy_info.yaml` specification is updated, now `gateway` field is removed and become otaclient internal configuration field derived from `upper_ota_proxy` field. 
For backward compatibility, with this PR, `gateway` field in the `proxy_info.yaml` will be ignored.

File: otaclient/configs/proxy_info.py (hunk 1 above)
Comment (Bodong-Yang) [reply otaclient/configs/proxy_info.py:144]: Now a logging containing the content of invalid yaml file will be issued when parsing invalid yaml file.

File: otaclient/configs/ecu_info.py [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+)
+
+
+def parse_ecu_info(ecu_info_file: StrOrPath) -> ECUInfo:
+    try:
+        _raw_yaml_str = Path(ecu_info_file).read_text()
+        loaded_ecu_info = yaml.safe_load(_raw_yaml_str)
+        assert isinstance(loaded_ecu_info, dict), "not a valid yaml file"
+        return ECUInfo.model_validate(loaded_ecu_info, strict=True)
```
[the remark this answers is not in the document]
Comment (Bodong-Yang) [reply otaclient/configs/ecu_info.py:140]: Now invalid ecu_info.yaml's content will be logged.

File: otaclient/configs/ecu_info.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (Bodong-Yang) [reply otaclient/configs/ecu_info.py:125]: `BootloaderType.UNSPECIFIED` now is renamed to `BootloaderType.AUTO_DETECT`. 
This bootloader type is only used when `bootloader` is not specified in `ecu_info.yaml`, and bootloader auto detection mechanism will be involved to try to detect the bootloader type.
`BootloaderType.AUTO_DETECT` is only meant for backward-compatibility and deprecated, users MUST specifies the correct bootloader type in the `ecu_info.yaml`.

File: tests/test_proxy_info.py [hunk 6]
```diff
@@ -13,107 +13,96 @@
...
+        (
+            "/t/t/t/t/t/t/t/tyaml file should not contain tabs/t/t/t/",
+            DEFAULT_PROXY_INFO,
+        ),
+        # ------ case 6: backward compatibility test ------ #
+        # for superseded field, the corresponding field should be assigned.
+        # for removed field, it should not impact the config file loading.
+        (
+            "enable_ota_proxy: true\ngateway: false\n",
+            DEFAULT_PROXY_INFO,
```
Comment (Bodong-Yang) [inline tests/test_proxy_info.py:99]: Test case 6 is added to ensure the backward-compatibility with old proxy_info.yaml.
1. `enable_ota_proxy` field's value will be assigned to new `enable_local_ota_proxy` field.
2. deprecated `gateway` field will be ignored when parsing the proxy_info.yaml.

**Pull request merged**